api/server/server.go

```diff
@@ -0,0 +1,160 @@
+package server
+
+import (
+	"context"
+	"io"
+	"net"
+	"sync"
+
+	"github.com/andres-teleport/overseer/api"
+	"github.com/andres-teleport/overseer/api/authentication"
+	"github.com/andres-teleport/overseer/lib/multipipe"
+	"github.com/andres-teleport/overseer/lib/supervisor"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	ErrEmptyCommand = status.Error(codes.InvalidArgument, "empty job command provided")
+)
+
+type Server struct {
+	jobOwners  map[string]string
+	mu         *sync.RWMutex
+	supervisor *supervisor.Supervisor
+	srv        *grpc.Server
+	l          net.Listener
+	api.UnimplementedJobworkerServiceServer
+}
+
+func NewServer(listenAddr, keyFile, certFile, caFile string) (*Server, error) {
+	creds, err := authentication.NewServerTransportCredentials(keyFile, certFile, caFile)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &Server{
+		jobOwners:  make(map[string]string),
+		mu:         &sync.RWMutex{},
+		supervisor: supervisor.NewSupervisor(),
+	}
+
+	authInterceptor := NewAuthorizationInterceptor(s)
+
+	s.srv = grpc.NewServer(
+		grpc.Creds(creds),
+		grpc.UnaryInterceptor(authInterceptor.unaryServerInterceptor),
+		grpc.StreamInterceptor(authInterceptor.streamServerInterceptor),
+	)
+	api.RegisterJobworkerServiceServer(s.srv, s)
+
+	l, err := net.Listen("tcp", listenAddr)
+	if err != nil {
+		return nil, err
+	}
+	s.l = l
+
+	return s, nil
+}
+
+func (s *Server) Serve() error {
+	return s.srv.Serve(s.l)
+}
+
+func (s *Server) Close() error {
+	return s.l.Close()
+}
+
+func (s *Server) Start(ctx context.Context, job *api.Job) (*api.JobID, error) {
+	commonName, err := authentication.GetCommonNameFromCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(job.Command) == 0 {
+		return nil, ErrEmptyCommand
+	}
+
+	jobID, err := s.supervisor.StartJob(job.Command, job.Arguments...)
+	if err != nil {
+		return nil, status.Error(codes.Aborted, err.Error())
+	}
+
+	s.mu.Lock()
+	s.jobOwners[jobID] = commonName
+	s.mu.Unlock()
+
+	resp := &api.JobID{Id: string(jobID)}
+
+	return resp, nil
+}
+
+func (s *Server) Stop(ctx context.Context, jobID *api.JobID) (*api.StopResponse, error) {
+	err := s.supervisor.StopJob(jobID.Id)
+
+	if err == supervisor.ErrJobFinished {
+		err = status.Error(codes.FailedPrecondition, err.Error())
+	} else {
+		err = status.Error(codes.Internal, err.Error())
+	}
```

Comment:
if err == nil the else statment will be invoked thus err == nil the `err.Error()` call in `err = status.Error(codes.Internal, err.Error())` will panic.